RocketDataTests/CollectionDataProviderTests/ConsistencyCollectionDataProviderTests.swift

```diff
@@ -229,6 +229,77 @@ class ConsistencyCollectionDataProviderTests: RocketDataTestCase {
         waitForExpectationsWithTimeout(10, handler: nil)
     }
 
+    /**
+     If we insert an item into a collection, possibly, other models in the collection should update.
+     So in this test, all the models have a submodel. We insert a new model which has an updated submodel.
+     This submodel should get updated everywhere in this collection.
+     
+     Specifically, we have two models initially. One has a submodel with id 2, the other without an id.
+     We add a new model to the array which has an updated submodel of id 2.
+     We expect the consistency manager to update this row later.
+     */
+    func testInsertUpdatesOtherItemsInCollections() {
+        let dataProvider = CollectionDataProvider<ParentModel>(dataModelManager: DataModelManager.sharedDataManagerNoCache)
+        let otherDataProvider = CollectionDataProvider<ParentModel>(dataModelManager: DataModelManager.sharedDataManagerNoCache)
+
+        let initialChildModel = ChildModel(id: 2, name: "initial")
+        let initialModel = ParentModel(id: 1, name: "initial", requiredChild: initialChildModel, otherChildren: [])
+        let otherModel = ParentModel(id: 3, name: "other", requiredChild: ChildModel(), otherChildren: [])
+        let newChild = ChildModel(id: 2, name: "new")
+        let newModel = ParentModel(id: 5, name: "new", requiredChild: newChild, otherChildren: [])
+
+        var delegateCalled = 0
+        let delegate = ClosureCollectionDataProviderDelegate() { (collectionChanges, context) in
+            delegateCalled += 1
+            switch delegateCalled {
+            case 1:
+                XCTAssertEqual(context as? String, "context")
+                XCTAssertEqual(dataProvider[2].name, "new")
+                XCTAssertEqual(collectionChanges.count, 1)
+                XCTAssertEqual(collectionChanges[0], CollectionChangeInformation.insert(index: 2))
+            case 2:
```

Comment:
Yes. This isn't the modified collection btw - this is the synced collection.

1 happens immediately when the insert happens. Both get the insert synchronously.
2 could be a lot of work (if it's deeply nested), so happens asyncronously through the consistency manager. I thought about making it syncronous but decided against it for a few reasons:

- Performance (note that the changed models are subtrees and they could be deeply nested)
- API. If this were the case, inserting a row could edit other rows. If you wanted to do tableview animations, this would be really tough. You'd have to return other index paths changed?
- Edge case. This shouldn't happen too often, and don't want to optimize for the edge case.
- Even though it's asynchronous, it should be really fast.

OK. So, next thing is: why doesn't this happen automatically.

For the first update, the data provider has a "not yet correct" collection (because other rows have not been updated). It gives this to the consistency manager. It's an object with id: CollectionID. It sees that this collection cares about this ID, so immediately just replaces the whole thing (but nothing's changed, so nothing gets updated). Basically, it doesn't look at the submodels because it doesn't need to. The top level id matches, so it can just replace the whole model.